### PR TITLE
Kevin-change notes to spawn by sample

### PR DESCRIPTION
**Pull request by KevinHuang78759**
## Summary
Loading in a level from Json now completely works
Music can also be played now
Note assets from the designers have been implemented
Notes now spawn by sample rather than frame

The way that Sami and I read Jsons differ slightly in that we may have differing file names and entry names. The overall structure should be relatively similar. This is because Sami was working mainly with the level editor whereas I was working  mainly with the game itself. Change and merge these as you see fit.

For the level to load properly, each BandMember's notes **MUST BE SORTED IN NON DECREASING ORDER BY START SAMPLE**

I would rather not use a priority queue and perform a bunch of unnecessary O(nlog n) heap sorts

## Notes 

Lane and background assets have been left for later because they are currently not in a form I am able to use.

I also have not accommodated for calibration results because it appears the specification for that is still a bit up in the air.

The infrastructure of band member competency is there, but it is not drawn. I would like to discuss with Gurvir on my foreseen implementation. 

We also still need to implement the win/lose condition.

### CODE REVIEW

File: assets/level.json
Comment (mluo24) [inline assets/level.json:1]: put this in a folder called `levels`

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java [hunk 1]
```diff
@@ -132,43 +145,119 @@ public void setDestroyed(boolean d){
...
-		}
-		y += vy;
+	/**
+	 * Set the textures for held notes. This is outside the constructor because not every note is a held note
+	 * @param trail
+	 * @param trailFrames
+	 * @param end
+	 * @param endFrames
+	 */
+	public void setHoldTextures(Texture trail, int trailFrames, Texture end, int endFrames){
```
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/entity/Note.java:226]: why don't you put this in the constructor? I think to solve the problem you bring up, you can just call it with an `if` statement or include it in the switch statement I feel like. that way if you initialize another held note somewhere you don't have to worry about this.

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java [hunk 2]
```diff
@@ -180,25 +269,47 @@ public void setTail_thickness(float t){
...
-			canvas.drawRect(x - tail_thickness/2, by, x + tail_thickness/2, y, Color.BLUE, true);
+			//The tail should be about half the width of the actual note assets
+			tail_thickness = widthConfine/2f;
+			//Set the animation frame properly
+			trailAnimator.setFrame((int)curTrailFrame);
+			endAnimator.setFrame((int)curEndFrame);
+			//Start at the bottom location, then draw until we reach the top
+			float cury = by;
+			//This for loop is cursed, but its really a while loop
+			for (;cury < y - trailHeight*tail_thickness/trailWidth; cury += trailHeight*tail_thickness/trailWidth){
```
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/entity/Note.java:290]: if your initialization does not happen in the for loop just make it a while loop lol cause this is indeed cursed

or just put the initialization in the for loop since I don't see you modifying it in the body anywhere

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java [hunk 3]
```diff
@@ -132,43 +145,119 @@ public void setDestroyed(boolean d){
...
+			case BEAT:
+				NUM_ANIM_FRAMES = 1;
+				break;
+			case HELD:
+				NUM_ANIM_FRAMES = 1;
+				break;
+			case SWITCH:
+				NUM_ANIM_FRAMES = 1;
+				break;
+		}
```
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/entity/Note.java:168]: why does this exist when they are all the same value

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java [hunk 4]
```diff
@@ -17,13 +17,21 @@
...
 	/** How fast we change frames (one frame per 4 calls to update) */
 	private static final float ANIMATION_SPEED = 0.25f;
+
+	public int getNUM_ANIM_FRAMES() {
+		return NUM_ANIM_FRAMES;
+	}
+
+	public void setNUM_ANIM_FRAMES(int NUM_ANIM_FRAMES) {
+		this.NUM_ANIM_FRAMES = NUM_ANIM_FRAMES;
+	}
```
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/entity/Note.java:31]: why is this necessary? I don't see you using it anywhere (and I don't think it should be able to be modified, even by the BandMember class)

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java [hunk 5]
```diff
@@ -44,20 +52,32 @@ public int getLine(){
...
-	public void setStartFrame(int t){
-		startFrame = t;
+
+	private long hitSample;
+	public long getHitSample(){
+		return hitSample;
+	}
+	public void setHitSample(long t){
+		hitSample = t;
 	}
```
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/entity/Note.java:69]: is this the sample at which the note gets hit?

**Review by mluo24 — COMMENTED** [review]
**NOTE I'M NOT DONE I accidentally pressed submit too early**

A lot of your `Note` code also makes me believe we should just make `Note` an abstract class or an interface (preferably an abstract class since a lot of logic with the notes are similar) and extend other classes from them, like `HitNote`, `SwitchNote`, and `HeldNote` because having everything in the `Note` class be mostly just `HeldNote` code being handled with enums feels weird to me. See `GameObject` as an example. 

level editor looks completely messed up so that should be fixed (i'm not sure what happened, it might be due to asset loading?)
image

Lastly, are we going to be using ms or samples? My argument for using ms is that doing an extra conversion to samples now no longer makes sense, since you're just multiplying by a constant that uses seconds as part of the calculation. we need to make this decision, like, before the presentation.

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java [hunk 6]
```diff
@@ -360,57 +324,10 @@ public void reset() {
...
-					Note n = new Note(l, Note.NoteType.HELD, frame, greenTexture);
-					n.setY(noteSpawnY);
-					n.setBottomY(noteSpawnY);
-					n.setHoldFrames(150);
-					bm.getAllNotes().addLast(n);
-				}
-			}
-		}
-	}
+	public int lpl = 4;
```
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java:330]: I don't think we need this anymore?

File: core/src/edu/cornell/gdiac/temporary/GameMode.java [hunk 7]
```diff
@@ -268,7 +244,7 @@ public void resize(int width, int height) {
 	public void render(float delta) {
 		if (active) {
 			update(delta);
-			draw(delta);
+			draw(gameplayController.level.getMusic().getCurrent().getStream().getSampleOffset());
 			if (inputController.didExit() && listener != null) {
```
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/GameMode.java:248]: I thought that this doesn't work?

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java [hunk 8]
```diff
@@ -546,18 +462,18 @@ public void handleActions(InputController input){
...
+		for(Note n : level.getBandMembers()[checkBM].getHitNotes()){
 			if(n.getNoteType() == Note.NoteType.BEAT){
 				if(triggers[n.getLine()] && !hitReg[n.getLine()]){
-					//Check for all the beats in this line and in the active band member
+					//Check for all the notes in this line and in the active band member
 					//See if any are close enough
-					float dist = Math.abs(hitbarY - n.getY())/n.getHeight();
-					if(dist < 1.5){
+					long dist = Math.abs(currentSample - n.getHitSample());
+					if(dist < 25000){
```
Comment (mluo24) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java:471]: just saying rn for every part that has logic with checking to see if it's on beat, this is where calibration comes into play. I really recommend separating this into a separate method so this logic doesn't get repeated or cluttered (like I did with `isOnBeat`. You don't have to worry about this because I will likely be working to refactor it, but like I said, ms or samples.

A lot of this seems unlabeled so it's slightly confusing.

**Review by mluo24 — APPROVED** [review]
sorry for double review, pressed submit too early for the first one

thanks for working on this! level class looks good, and I think a lot of stuff here that would need to be changed is based on future changes to the arch spec, which are going to happen this week.

variable naming and conventions can still use some improvement as usual.

I'm going to approve it because I don't see any glaring issues (at the moment, except for the level editor), so just keep my comments in mind and address them. And fix the level editor.

File: assets/level.json
Comment (Sami362002) [inline assets/level.json:1]: I don't think having a connections list of two objects for every held note is a good way to do it. Just give the held note a location like a normal note and then a duration in the same time units.

File: core/src/edu/cornell/gdiac/temporary/GameMode.java [hunk 9]
```diff
@@ -61,43 +69,40 @@ public enum GameState {
...
+		gameplayController = new GameplayController(canvas.getWidth(),canvas.getHeight());
+
+
+	}
+
+	public void readLevel(AssetDirectory directory){
+		JsonReader jr = new JsonReader();
+		JsonValue levelData = jr.parse(Gdx.files.internal("level.json"));
+		gameplayController.loadLevel(levelData, directory);
+		inputController = new InputController(gameplayController.NUM_LANES, gameplayController.lpl);
```
Comment (Sami362002) [inline core/src/edu/cornell/gdiac/temporary/GameMode.java:105]: Remember that everyone, including my editor, the menu, and calibration, is using the InputController, so it shouldn't take in the number of lanes and lines as a constructor. Instead I think you should have a function setLanes() which you call to et the correct amount of lanes, only then can you call the other input methods which require knowing the lane amount.

File: assets/level.json [hunk 10]
```diff
@@ -0,0 +1,83 @@
+{
```
Comment (Sami362002) [inline assets/level.json:1]: I don't think having a connections list of two objects for every held note is a good way to do it. Just give the held note a location like a normal note and then a duration in the same time units.

File: core/src/edu/cornell/gdiac/temporary/GameMode.java [hunk 11]
```diff
@@ -268,7 +244,7 @@ public void resize(int width, int height) {
 	public void render(float delta) {
 		if (active) {
 			update(delta);
-			draw(delta);
+			draw(gameplayController.level.getMusic().getCurrent().getStream().getSampleOffset());
```
Comment (Sami362002) [inline core/src/edu/cornell/gdiac/temporary/GameMode.java:247]: you probably realized it already, but this doesn't work because getStream() gets the byte stream of the audio but methods from the Music class don't actually update the audio stream. As a results, the current sample remains at zero. Instead you gotta get the time from the Music class with getPosition(). This isn't as precise but whatevs. It also gives the time in seconds. For my level editor, because I made everything in samples, I just converted with the simple formula: SampleOffset = sampleRate*getPosition();

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java [hunk 12]
```diff
@@ -158,6 +151,17 @@ public GameplayController(int lanes, int linesPerLane, float width, float height
...
 		TOPBOUND = 19f*height/20f;
 		BOTTOMBOUND = height/5f;
+
+
+	}
+
+	/**
+	 * Loads a level
+	 */
+	public void loadLevel(JsonValue levelData, AssetDirectory directory){
```
Comment (Sami362002) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java:161]: I think both you and me ran into an annoying problem (im just assuming here) where you have to initialize the gameplay controller before you can give it the level json or the directory. As a result, you do a bit of initialization in the gameController constructor, and then once you feed in the directory, you gotta call another function (this function) to get all the level details and finish initialization. It's annoying but idk whatever i guess.

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java [hunk 13]
```diff
@@ -169,42 +173,27 @@ public GameplayController(int lanes, int linesPerLane, float width, float height
...
-			bandMembers[i].setHeight(TOPBOUND - BOTTOMBOUND);
-			bandMembers[i].setNumLines(lpl);
-			bandMembers[i].setMaxComp(MAX_HEALTH);
-			bandMembers[i].setCurComp(MAX_HEALTH);
-			XCoor += bandMembers[i].getWidth() + inBetweenWidth;
-		}
+	 */
+	public void setupBandMembers(){
+		level.setActiveProperties(0, largewidth, smallwidth, TOPBOUND - BOTTOMBOUND);
+		level.setBandMemberBl(new Vector2(LEFTBOUND, BOTTOMBOUND), inBetweenWidth);
```
Comment (Sami362002) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java:196]: I cannot understand what BL refers to in this method name. I don't think two letter abbreviations usually work well.

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java [hunk 14]
```diff
@@ -213,19 +202,35 @@ public void setupBandMembers(Color[] c){
 	 * destroyed notes is also done in here.
 	 */
 	public void checkDeadNotes(){
-		for(int i = 0; i < bandMembers.length; ++i){
-			for(Note n : bandMembers[i].getHitNotes()){
+		for(int i = 0; i < level.getBandMembers().length; ++i){
```
Comment (Sami362002) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java:205]: you could just loop through the band members themselves with [for (BandMember member : bandMembers)] to make the code more concise. When you check if [ i == activeBM], you could just check if that member itself is the active one (im assuming you have a field for that in the band member class...)

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java (hunk 8 above)
Comment (Sami362002) [reply core/src/edu/cornell/gdiac/temporary/GameplayController.java:471]: I think the calibration offset will be factored into the n.getHitSample() value

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java [hunk 15]
```diff
@@ -546,18 +462,18 @@ public void handleActions(InputController input){
...
-		for(Note n : bandMembers[checkBM].getHitNotes()){
+		for(Note n : level.getBandMembers()[checkBM].getHitNotes()){
 			if(n.getNoteType() == Note.NoteType.BEAT){
 				if(triggers[n.getLine()] && !hitReg[n.getLine()]){
-					//Check for all the beats in this line and in the active band member
+					//Check for all the notes in this line and in the active band member
 					//See if any are close enough
-					float dist = Math.abs(hitbarY - n.getY())/n.getHeight();
-					if(dist < 1.5){
+					long dist = Math.abs(currentSample - n.getHitSample());
```
Comment (Sami362002) [inline core/src/edu/cornell/gdiac/temporary/GameplayController.java:470]: ew longs. Just work with ints

File: core/src/edu/cornell/gdiac/temporary/Level.java [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+   private MusicQueue music;
+
+    /**
+     * The last sample that health was decremented due to continuous decay
+     */
+    private long lastDec;
+    public Level(JsonValue data, AssetDirectory directory) {
+        //Read in Json  Value and populate asset textures
+        lastDec = 0;
+        title = data.getString("title");
```
Comment (Sami362002) [inline core/src/edu/cornell/gdiac/temporary/Level.java:108]: wait did you change the JSON format? I am the one who was supposed to be deciding the JSON format you can't just change it. My level editor saves the level into the format I specified. Most of the differences are names of the fields, but for held notes I don't have a "connections" field, I have a "sample" field which I call "position" and "duration". You gotta change that before we merge.

File: core/src/edu/cornell/gdiac/temporary/editor/EditorMode.java [hunk 17]
```diff
@@ -372,7 +372,7 @@ public void dispose() {
     public void populate(AssetDirectory directory) {
         background  = directory.getEntry("background",Texture.class);
         displayFont = directory.getEntry("times",BitmapFont.class);
-        catNoteTexture = directory.getEntry("catnote", Texture.class);
+        catNoteTexture = directory.getEntry("hit", Texture.class);
```
Comment (Sami362002) [inline core/src/edu/cornell/gdiac/temporary/editor/EditorMode.java]: weird name but ok

File: core/src/edu/cornell/gdiac/temporary/entity/Note.java (hunk 1 above)
Comment (Sami362002) [reply core/src/edu/cornell/gdiac/temporary/entity/Note.java:226]: This is how they set textures in the lab. I think it's because you need to wait until populate() is ran to actually get access to the textures. And that only happens after initialization of the notes

**Review by Sami362002 — APPROVED** [review]
I liked your documentation 😸

File: core/src/edu/cornell/gdiac/temporary/GameplayController.java (hunk 8 above)
Comment (mluo24) [reply core/src/edu/cornell/gdiac/temporary/GameplayController.java:471]: it is yeah
